ros2_ws/Dockerfile

```diff
@@ -137,10 +144,45 @@ WORKDIR ${ROS2_WORKSPACE}
 RUN /bin/bash -c "source /opt/ros/${ROS_DISTRO}/setup.bash; colcon build"
 RUN rm -rf ./src
 
-FROM ros2-control-${ROS_DISTRO} as ros2-control
+FROM base-workspace as ros2-control-jazzy
+RUN apt-get update && apt-get install --no-install-recommends -y \
+  # doxygen \
+  # librange-v3-dev \
+  python3-jinja2 \
+  python3-typeguard \
+  && rm -rf /var/lib/apt/lists/*
+
+# build as ROS user
+USER ${USER}
+WORKDIR ${ROS2_WORKSPACE}/src
+
+RUN git clone -b 4.11.0 --depth 1 https://github.com/ros-controls/ros2_control.git
+RUN git clone -b 4.8.0 --depth 1 https://github.com/ros-controls/ros2_controllers.git
+# get additional interface dependencies manually
+RUN git clone -b 5.1.0 --depth 1 https://github.com/ros-controls/control_msgs.git
+RUN git clone -b 1.0.5 --depth 1 https://github.com/pal-robotics/backward_ros.git
+RUN git clone -b 4.0.0 --depth 1 https://github.com/ros/diagnostics.git
+RUN git clone -b 2.5.0 --depth 1 https://github.com/ros-controls/realtime_tools.git
+RUN git clone -b 2.1.0 --depth 1 https://github.com/ros2/rcl_interfaces.git
+RUN git clone -b 0.12.0 --depth 1 https://github.com/ros2/test_interface_files.git
+RUN git clone -b 1.16.0 --depth 1 https://github.com/ros/angles.git
+RUN git clone -b ros2 --depth 1 https://github.com/ros-drivers/ackermann_msgs.git
+RUN git clone -b 0.3.8 --depth 1 https://github.com/PickNikRobotics/generate_parameter_library.git
+RUN git clone -b 1.1.0 --depth 1 https://github.com/PickNikRobotics/RSL.git
+RUN git clone -b 1.0.2 --depth 1 https://github.com/PickNikRobotics/cpp_polyfills.git
+RUN git clone -b 3.2.0 --depth 1 https://github.com/ros-controls/control_toolbox.git
+RUN git clone -b 2.1.2 --depth 1 https://github.com/ros/filters.git
+RUN git clone -b 1.1.0 --depth 1 https://github.com/ros-controls/kinematics_interface.git
+
+WORKDIR ${ROS2_WORKSPACE}
+RUN /bin/bash -c "source /opt/ros/${ROS_DISTRO}/setup.bash; colcon build"
+RUN rm -rf ./src
+
+FROM ros2-control-${BASE_TAG} as ros2-control
 
 FROM environment-variables as final
 COPY --from=ros2-control / /
+RUN rm -rf /home/ubuntu
```

Comment:
A slightly annoying nitpick/request, but would it make more sense to move this command directly after renaming the user?

Reply:
You cannot, it's not there after renaming the user. We do `FROM environment-variables as final` and that directory is actually coming from the `environment-variables` stage where we can't delete it yet since the user is still called ubuntu at this point.

The reason for these 2 lines (183 and 184) is that we want to reduce the amount of layers in the final image. Every `RUN` instruction creates a layer and an image can only have so many layers. That's why it's good practice to reduce layers, one way to do that is to do something like in these two lines, start again from the base image and just copy paste the content of the previous stage into the base. The fact that we then have this directory again is a small side effect of that trick.

Reply:
oh right, I see. Ok, then this sounds good to me. Thanks!